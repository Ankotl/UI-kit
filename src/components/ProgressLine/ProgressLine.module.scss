@import '../../styles/tokens/_colors';
@import '../../styles/tokens/_transitions';

.ProgressLine__line {
  border-radius: 10px;
}

.ProgressLine__mainLine {
  position: relative;
  overflow: hidden;
  height: 6px;
  background-color: $color-light-grey;
}

.ProgressLine__secondaryLine {
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  background-color: $color-green;
  transition: width $transition-main;
}