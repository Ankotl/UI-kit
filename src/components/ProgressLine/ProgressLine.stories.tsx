import React from 'react'

import { getStoryMeta } from '../../utils/storybook/getStoryMeta'
import { getStoryTemplate } from '../../utils/storybook/getStoryTemplate'

import {
  IProgressLineProps,
  ProgressLine as ProgressLineCmp,
} from './ProgressLine'

const storiesMeta = getStoryMeta({
  component: ProgressLineCmp,
})
export default storiesMeta

const storyTemplate = getStoryTemplate((props: IProgressLineProps) => {
  const containerStyle = {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    height: '100vh',
  }

  return (
    <div style={containerStyle}>
      <ProgressLineCmp {...props} />
    </div>
  )
})

export const ProgressLine = storyTemplate({
  fixedPercentage: 0,
})
