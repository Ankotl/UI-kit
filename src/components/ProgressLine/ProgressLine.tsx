import React from 'react'
import { SpringValue, animated } from 'react-spring'

import clsx from 'clsx'

import s from './ProgressLine.module.scss'

export interface IProgressLineClasses {
  mainLineClass?: string
  secondaryLineClass?: string
  lineClass?: string
}

export interface IProgressLineBaseProps {
  classes?: IProgressLineClasses
  className?: string
}

export interface IProgressLineFixedProps {
  fixedPercentage: number
}

export interface IProgressLineAnimatedProps {
  animatedStyle: Record<string, SpringValue<any>>
}

export type IProgressLineProps = IProgressLineBaseProps &
  (IProgressLineFixedProps | IProgressLineAnimatedProps)

const isAnimatedComponent = (
  props: IProgressLineProps
): props is IProgressLineAnimatedProps => {
  return 'animatedStyle' in props
}

export const ProgressLine: React.FC<IProgressLineProps> = ({
  className,
  classes,
  ...otherProps
}) => {
  const renderSecondLine = () => {
    const _className = clsx(
      s.ProgressLine__line,
      s.ProgressLine__secondaryLine,
      classes?.secondaryLineClass,
      classes?.lineClass
    )

    if (isAnimatedComponent(otherProps)) {
      return (
        <animated.div className={_className} style={otherProps.animatedStyle} />
      )
    }

    if (!isAnimatedComponent(otherProps)) {
      return (
        <div
          className={_className}
          style={{ width: `${otherProps.fixedPercentage}%` }}
        />
      )
    }

    return null
  }

  return (
    <div
      className={clsx(
        s.ProgressLine__line,
        s.ProgressLine__mainLine,
        classes?.mainLineClass,
        classes?.lineClass,
        className
      )}
    >
      {renderSecondLine()}
    </div>
  )
}
