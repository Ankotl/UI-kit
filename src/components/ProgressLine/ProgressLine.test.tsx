import React from 'react'

import { render } from '@testing-library/react'

import { ProgressLine } from './ProgressLine'

describe('ProgressLine component test', () => {
  test('Has right states', () => {
    const { container } = render(<ProgressLine fixedPercentage={80} />)
    expect(container).toMatchSnapshot()
  })
})
