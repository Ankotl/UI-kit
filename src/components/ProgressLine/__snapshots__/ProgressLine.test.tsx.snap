// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressLine component test Has right states 1`] = `
<div>
  <div
    class="ProgressLine__line ProgressLine__mainLine"
  >
    <div
      class="ProgressLine__line ProgressLine__secondaryLine"
      style="width: 80%;"
    />
  </div>
</div>
`;
